libmscore/scale.cpp

```diff
@@ -0,0 +1,265 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2011-2012 Werner Schweer and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2
+//  as published by the Free Software Foundation and appearing in
+//  the file LICENSE.GPL
+//=============================================================================
+
+#include "scale.h"
+#include "mscore.h"
+#include "note.h"
+#include <set>
+
+using namespace std;
+
+namespace Ms {
+
+//---------------------------------------------------------
+//   getNextLine
+//---------------------------------------------------------
+
+QString Scale::getNextLine(QTextStream& in)
+      {
+      while (!in.atEnd()) {
+            QString line = in.readLine();
+            if (line[0] == '!')
+                  continue;
+            return line;
+            }
+      return QString();
+      }
+
+//---------------------------------------------------------
+//   loadScale
+//---------------------------------------------------------
+
+bool Scale::loadScale(const QString& fn)
+      {
+      static set<int> supportedNbs = { 8, 12, 21, 35 };
+
+      QFile f(fn);
+      if (!f.open(QIODevice::ReadOnly)) {
+            MScore::lastError = strerror(errno);
+            return false;
+            }
+
+      QTextStream in(&f);
+
+      // Read description of the scala file
+      QString description = getNextLine(in);
+
+      int nbNotes = getNextLine(in).toInt();
+      if (supportedNbs.find(nbNotes + 1) == supportedNbs.end()) {
+            MScore::lastError = QObject::tr("MuseScore supports only Scala files with 8, 12, 21 or 35 notes");
+            return false;
+            }
+
+      for (int i = 0; i < nbNotes; ++i) {
+            QString pitchLine = getNextLine(in);
+            if (pitchLine.contains('/')) {
+                  QStringList values = pitchLine.split('/');
+                  if (values.size() != 2 || values[1].toDouble() == 0) {
+                        MScore::lastError = QObject::tr("Unsupported Scala file format.");
+                        return false;
+                        }
+
+                  originalNotes.push_back(1200 * (values[0].toDouble() / values[1].toDouble()) - 1200);
```

Comment:
Not sure this is correct. If the value is expressed as a ratio, it is a linear value, not proportional to any cent value, which are logarithmic.

What you need is something like this:

`value = log(values[0].toDouble() / values[1].toDouble()) / log(pow(2, 1/1200));`

where `log(pow(2, 1/1200))` is the natural logarithm of 1 cent and can be pre-computed once for all.